### PR TITLE
[CARBONDATA-4285] Fix alter add complex columns with global sort compaction failure

### PR DESCRIPTION
**Pull request by maheshrajus**
### Why is this PR needed?
 Alter add complex columns with global sort compaction is failing due to 
1) AOI exception : Currently creating default complex delimiter list in global sort compaction with size of 3. For map case need extra complex delimiter for handling the key-value
2) bad record handling: When we  add complex columns after insert the data, complex columns has null data for previously loaded segments. this null value is going to treat as bad record and compaction is failed.
 
 ### What changes were proposed in this PR?
1) In Global sort compaction flow create default complex delimiter with 4, as already doing in load flow. 
2)  Bad records handling pruned for compaction case. No need to check bad records for compaction as they are already checked while loading. previously loaded segments data we are inserting again in compaction case
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/319/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5917/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4173/

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java [hunk 1]
```diff
@@ -412,16 +412,18 @@ private void updateValueToByteStream(DataOutputStream dataOutputStream, byte[] v
...
-      message = CarbonDataProcessorUtil
-          .prepareFailureReason(carbonDimension.getColName(), carbonDimension.getDataType());
-      logHolder.getColumnMessageMap().put(carbonDimension.getColName(), message);
+    if (null == input) {
+      String message = logHolder.getColumnMessageMap().get(carbonDimension.getColName());
+      if (null == message) {
+        message = CarbonDataProcessorUtil
+            .prepareFailureReason(carbonDimension.getColName(), carbonDimension.getDataType());
+        logHolder.getColumnMessageMap().put(carbonDimension.getColName(), message);
+        logHolder.setReason(message);
```
Comment (ShreelekhyaG) [inline processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java:424]: can move setReason to line 426 after closing one if case of (null == message)

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java [hunk 2]
```diff
@@ -171,13 +171,13 @@ public boolean getIsColumnDictionary() {
   }
 
   @Override
-  public void writeByteArray(ArrayObject input, DataOutputStream dataOutputStream,
+  public void writeByteArray(Object input, DataOutputStream dataOutputStream,
       BadRecordLogHolder logHolder, Boolean isWithoutConverter) throws IOException {
-    if (input == null) {
+    if (input == null || input.equals("\\N")) {
```
Comment (ShreelekhyaG) [inline processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java:176]: check if we could get `SERIALIZATION_NULL_FORMAT ` while creating complex type from FieldEncoderFactory

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala [hunk 3]
```diff
@@ -583,4 +583,78 @@ class TestAlterTableAddColumns extends QueryTest with BeforeAndAfterAll {
...
+    checkAnswer(sql("select * from alter_global1"),
+      Seq(Row(1, null), Row(2, null), Row(3, null), Row(4, make(Array(1)))))
+    val addedColumns = addedColumnsInSchemaEvolutionEntry("alter_global1")
+    assert(addedColumns.size == 1)
+    sql("alter table alter_global1 compact 'minor'")
+    checkAnswer(sql("select * from alter_global1"),
+      Seq(Row(1, null), Row(2, null), Row(3, null), Row(4, make(Array(1)))))
+    sql("DROP TABLE IF EXISTS alter_global1")
+  }
+
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala:605]: remove extra line here

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala [hunk 4]
```diff
@@ -583,4 +583,78 @@ class TestAlterTableAddColumns extends QueryTest with BeforeAndAfterAll {
...
+        Row(1, make(Array(make(Array(1, 2)))), make(Array(Row("st", Map("a" -> "b")))),
+          Row("hi", make(Array(1, 2))), Row(2.3, Map("ph" -> make(Array(1, 2)))), null, null),
+        Row(1, make(Array(make(Array(1, 2)))), make(Array(Row("st", Map("a" -> "b")))),
+          Row("hi", make(Array(1, 2))), Row(2.3, Map("ph" -> make(Array(1, 2)))),
+          Map("a" -> make(Array("hi"))), Map("a" -> Row(23, Row("sh"))))
+      ))
+    val addedColumns = addedColumnsInSchemaEvolutionEntry("alter_global2")
+    assert(addedColumns.size == 6)
+    sql("alter table alter_global2 compact 'minor'")
+    checkAnswer(sql("select * from alter_global2"),
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala:648]: check answer can be reused here

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala [hunk 5]
```diff
@@ -583,4 +583,78 @@ class TestAlterTableAddColumns extends QueryTest with BeforeAndAfterAll {
...
+        " struct2 struct<num:double,contact:map<string,array<int>>>) ")
+    sql("insert into alter_global2 values(1, " +
+        "array(array(1,2)), array(named_struct('a1','st','map1', map('a','b'))), " +
+        "named_struct('s1','hi','arr',array(1,2)), named_struct('num',2.3,'contact',map('ph'," +
+        "array(1,2))))")
+    // multi-level nested map
+    sql(
+      "ALTER TABLE alter_global2 ADD COLUMNS(map1 map<string,array<string>>, map2 map<string," +
+      "struct<d:int, s:struct<im:string>>>)")
+    sql("insert into alter_global2 values(1,  " +
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala:630]: please add a insert with null and empty data also

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java (hunk 1 above)
Comment (maheshrajus) [reply processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java:424]: done

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala (hunk 3 above)
Comment (maheshrajus) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala:605]: done

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/323/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5921/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4176/

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala (hunk 4 above)
Comment (maheshrajus) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala:648]: done

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala (hunk 5 above)
Comment (maheshrajus) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/alterTable/TestAlterTableAddColumns.scala:630]: done

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java (hunk 2 above)
Comment (maheshrajus) [reply processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java:176]: done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4178/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5923/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/325/

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java [hunk 6]
```diff
@@ -100,13 +102,15 @@ public ArrayDataType(String name, String parentName, String columnId) {
    * @param parentName
    * @param columnId
    * @param isDictionary
+   * @param nullFormat
    */
   public ArrayDataType(String name, String parentName, String columnId,
-      Boolean isDictionary) {
+      Boolean isDictionary, String nullFormat) {
```
Comment (Indhumathi27) [inline processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java:108]: We no need to handle this way. Here, compaction is because of logging bad record error for inserting NULL for previous segments. Bad record should be done for Load/Insert flow only. We can avoid performing Bad record validation during validation by passing a flag from global sort flow. Please revert these changes, as it might induce new issues if actual data is NULL/empty data type

Comment (maheshrajus) [reply processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java:108]: done

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala [hunk 7]
```diff
@@ -342,7 +347,12 @@ object DataLoadProcessBuilderOnSpark {
       .mapPartitionsWithIndex { case (index, rows) =>
         ThreadLocalSessionInfo.setConfigurationToCurrentThread(conf.value.value)
         DataLoadProcessorStepOnSpark
-          .convertFunc(rows, index, modelBroadcast, partialSuccessAccum, convertStepRowCounter)
+          .convertFunc(rows,
+            index,
+            modelBroadcast,
+            true,
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala:353]: this flow looks like load flow. This change might not be needed. can test and remove if not required

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4180/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5925/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala (hunk 7 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala:353]: done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/328/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5926/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4181/

**Comment by Indhumathi27** [comment]
LGTM

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala [hunk 8]
```diff
@@ -219,6 +219,7 @@ object DataLoadProcessorStepOnSpark {
       rows: Iterator[CarbonRow],
       index: Int,
       modelBroadcast: Broadcast[CarbonLoadModel],
+      isCompactionFlow: Boolean = false,
```
Comment (kunal642) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala:222]: why add this in between? if you keep in last then scala will take the default value if not passed, so change in DataLoadProcessBuilderOnSpark.scala will not be required

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala [hunk 9]
```diff
@@ -342,7 +347,12 @@ object DataLoadProcessBuilderOnSpark {
       .mapPartitionsWithIndex { case (index, rows) =>
         ThreadLocalSessionInfo.setConfigurationToCurrentThread(conf.value.value)
         DataLoadProcessorStepOnSpark
-          .convertFunc(rows, index, modelBroadcast, partialSuccessAccum, convertStepRowCounter)
+          .convertFunc(rows,
+            index,
+            modelBroadcast,
+            false,
```
Comment (kunal642) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala:353]: not required to pass after making isCompactionFlow as the last argument

**Comment by ShreelekhyaG** [comment]
LGTM

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala (hunk 8 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala]: done

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala [hunk 10]
```diff
@@ -221,13 +221,15 @@ object DataLoadProcessorStepOnSpark {
       modelBroadcast: Broadcast[CarbonLoadModel],
       partialSuccessAccum: LongAccumulator,
       rowCounter: LongAccumulator,
-      keepActualData: Boolean = false): Iterator[CarbonRow] = {
+      keepActualData: Boolean = false,
+      isCompactionFlow: Boolean = false): Iterator[CarbonRow] = {
     val model: CarbonLoadModel = modelBroadcast.value.getCopyWithTaskNo(index.toString)
     val conf = DataLoadProcessBuilder.createConfiguration(model)
     val badRecordLogger = BadRecordsLoggerProvider.createBadRecordLogger(conf)
```
Comment (kunal642) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala:228]: instead of adding get/set in badrecord logger we can pass isCompactionFlow to BadRecordsLoggerProvider.createBadRecordLogger and disable the BadRecordLogger in that method.

That will be a clean implementation

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/329/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala (hunk 10 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessorStepOnSpark.scala:228]: OK, Done

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala (hunk 9 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala]: done

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4182/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5927/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala [hunk 11]
```diff
@@ -342,7 +348,11 @@ object DataLoadProcessBuilderOnSpark {
       .mapPartitionsWithIndex { case (index, rows) =>
         ThreadLocalSessionInfo.setConfigurationToCurrentThread(conf.value.value)
         DataLoadProcessorStepOnSpark
-          .convertFunc(rows, index, modelBroadcast, partialSuccessAccum, convertStepRowCounter)
+          .convertFunc(rows,
+            index,
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/load/DataLoadProcessBuilderOnSpark.scala:352]: please revert this change

File: processing/src/main/java/org/apache/carbondata/processing/loading/BadRecordsLoggerProvider.java [hunk 12]
```diff
@@ -33,6 +33,16 @@
...
+    return createBadRecordLogger(configuration, false);
+  }
+  /**
+   * method returns the BadRecordsLogger instance
+   * @param configuration
+   * @return
+   */
+
+  public static BadRecordsLogger createBadRecordLogger(CarbonDataLoadConfiguration configuration,
+      Boolean isForceDisale) {
```
Comment (Indhumathi27) [inline processing/src/main/java/org/apache/carbondata/processing/loading/BadRecordsLoggerProvider.java:45]:
```suggestion
      Boolean isForceDisable) {
```

File: processing/src/main/java/org/apache/carbondata/processing/loading/BadRecordsLoggerProvider.java [hunk 13]
```diff
@@ -33,6 +33,16 @@
...
    */
   public static BadRecordsLogger createBadRecordLogger(CarbonDataLoadConfiguration configuration) {
+    return createBadRecordLogger(configuration, false);
+  }
+  /**
+   * method returns the BadRecordsLogger instance
+   * @param configuration
+   * @return
+   */
+
```
Comment (ShreelekhyaG) [inline processing/src/main/java/org/apache/carbondata/processing/loading/BadRecordsLoggerProvider.java:43]: remove extra line at 43 and add one at 38

File: processing/src/main/java/org/apache/carbondata/processing/loading/BadRecordsLoggerProvider.java [hunk 14]
```diff
@@ -33,6 +33,16 @@
    * @return
    */
   public static BadRecordsLogger createBadRecordLogger(CarbonDataLoadConfiguration configuration) {
+    return createBadRecordLogger(configuration, false);
+  }
+  /**
```
Comment (Indhumathi27) [inline processing/src/main/java/org/apache/carbondata/processing/loading/BadRecordsLoggerProvider.java:38]: please format this comment. empty line should be added at line:38 and removed in line:43

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5930/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4185/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/332/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/338/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4191/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5937/

**Comment by kunal642** [comment]
LGTM..PR needs some refactoring..please fix in another PR
